esp-hal/src/dma/m2m.rs

```diff
@@ -31,145 +40,467 @@ pub struct Mem2Mem<'d, Dm>
 where
     Dm: DriverMode,
 {
-    channel: Channel<Dm, AnyGdmaChannel<'d>>,
-    rx_chain: DescriptorChain,
-    tx_chain: DescriptorChain,
-    peripheral: DmaPeripheral,
+    /// RX Half
+    pub rx: Mem2MemRx<'d, Dm>,
+    /// TX Half
+    pub tx: Mem2MemTx<'d, Dm>,
 }
 
 impl<'d> Mem2Mem<'d, Blocking> {
     /// Create a new Mem2Mem instance.
     pub fn new(
         channel: impl DmaChannelConvert<AnyGdmaChannel<'d>>,
         peripheral: impl DmaEligible,
-        rx_descriptors: &'static mut [DmaDescriptor],
-        tx_descriptors: &'static mut [DmaDescriptor],
-    ) -> Result<Self, DmaError> {
-        unsafe {
-            Self::new_unsafe(
-                channel,
-                peripheral.dma_peripheral(),
-                rx_descriptors,
-                tx_descriptors,
-                crate::dma::CHUNK_SIZE,
-            )
-        }
-    }
-
-    /// Create a new Mem2Mem instance with specific chunk size.
-    pub fn new_with_chunk_size(
-        channel: impl DmaChannelConvert<AnyGdmaChannel<'d>>,
-        peripheral: impl DmaEligible,
-        rx_descriptors: &'static mut [DmaDescriptor],
-        tx_descriptors: &'static mut [DmaDescriptor],
-        chunk_size: usize,
-    ) -> Result<Self, DmaError> {
-        unsafe {
-            Self::new_unsafe(
-                channel,
-                peripheral.dma_peripheral(),
-                rx_descriptors,
-                tx_descriptors,
-                chunk_size,
-            )
-        }
+    ) -> Self {
+        unsafe { Self::new_unsafe(channel, peripheral.dma_peripheral()) }
     }
 
     /// Create a new Mem2Mem instance.
     ///
     /// # Safety
     ///
-    /// You must ensure that your not using DMA for the same peripheral and
-    /// that your the only one using the DmaPeripheral.
+    /// You must ensure that you're not using DMA for the same peripheral and
+    /// that you're the only one using the DmaPeripheral.
     pub unsafe fn new_unsafe(
         channel: impl DmaChannelConvert<AnyGdmaChannel<'d>>,
         peripheral: DmaPeripheral,
```

Comment:
we should really take a compatible peripheral singleton here, we could avoid this being unsafe then. we have the mem2mem peripherals defined

Reply:
That's what the safe constructor does.

The unsafe one allows you to (for example) use SPI2 (without DMA) in the SPI Driver whilst using the "SPI2 trigger" for DMA memcpy at the same time.

This is an S3 and C6 problem which was fixed in later chips.